### DISCUSSION TITLE
Custom preview domain sets NEXT_PUBLIC_VERCEL_ENV to 'production'

### DISCUSSION THREAD

**Discussion by karmaniverous** [vercel/community#775] [Show & Tell]
I opened a ticket on this one but there doesn't seem to be a way to link back to it, so I thought I would also add a discussion here. Here's [a full writeup](https://karmanivero.us/blog/vercel-preview-branch-weirdness/) of the issue along with my workaround.

**Comment by amyegan** [comment]
Thanks for sharing this writeup @karmaniverous, and I'm sorry you're running into this problem. I was not able to reproduce the issue in my own project with two different custom domains (one using  CNAME Record and the other using Vercel nameservers). I have a couple of initial ideas, though you may have already thought of them:

- Are you using [Vercel CLI](https://vercel.com/docs/cli) or the [GitHub integration](https://vercel.com/docs/concepts/git/vercel-for-github) to deploy your preview branch? Wondering if there could be some kind of build difference there, and I want to make sure I'm replicating this as closely as possible to your own process. 
- Have you tried logging `NEXT_PUBLIC_VERCEL_ENV` and `VERCEL_ENV` to see if there are any differences in the values coming through? Might be good to try also logging `NEXT_PUBLIC_COMING_SOON` to rule out any glitch with that value.

If anyone else has run into this problem with preview environment variables, I hope they will share their experience here as well! This is not something I have seen before. Curious to understand what is happening to cause this behavior.

**Comment by charleskoehl** [comment]
We had this problem too. We're not using [Vercel CLI](https://vercel.com/docs/cli) or the [GitHub integration](https://vercel.com/docs/concepts/git/vercel-for-github) to deploy to our preview branch, which we have named "staging."

We want to use process.env.VERCEL_ENV while rendering a custom _document.js in our next.js app in order to add a noindex,nofollow meta tag to the <head> tag; however, process.env.VERCEL_ENV was always 'production' in both preview and production. (We added a comment to our inline google-tag-manager script to show the value, seen below in the html snippet.)

Now here's the weird part: Our staging subdomain was "signup.staging" when we observed process.env.VERCEL_ENV always being "production." Then on a hunch, we changed the subdomain to "signup-staging" so there were only 3 segments to the hostname (for example, "signup-staging.example.com" For some reason, this change fixed the issue. Now process.env.VERCEL_ENV === 'preview' at signup-staging.example.com and "production" at signup.example.com

```html
export default class MyDocument extends Document {
  render() {
    return (
      <Html lang='en'>
        <Head>
          {process.env.VERCEL_ENV !== 'production' && (
            <meta name='robots' content='noindex,nofollow' />
          )}
          <Script id='google-tag-manager' strategy='beforeInteractive'>
            {`
              (function(w,d,s,l,i){w[l]=w[l]||[];w[l].push({'gtm.start':
              new Date().getTime(),event:'gtm.js'});var f=d.getElementsByTagName(s)[0],
              j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';j.async=true;j.src=
              'https://www.googletagmanager.com/gtm.js?id='+i+dl;f.parentNode.insertBefore(j,f);
              })(window,document,'script','dataLayer','GTM-XXXXXX');
              window.dataLayer = window.dataLayer || [];
              function gtag(){dataLayer.push(arguments);}
              gtag('js', new Date());
              gtag('config', 'UA-99999999-1', {'transport_type': 'image'});
              // VERCEL_ENV = "${process.env.VERCEL_ENV}"
            `}
          </Script>
```

**Reply by charleskoehl** [reply]
Sorry, but we misread the VERCEL_ENV value on production. Now VERCEL_ENV === "preview" on both preview and production.

We've verified we have the same code on our main and staging branches. The signup-staging.example.com is assigned to staging branch, and signup.example.com is assigned to main branch.

Again, we're just pushing to staging and main to deploy.

**Reply by amyegan** [reply]
Hi @charleskoehl! Since you said you're not using Vercel CLI or the GitHub integration to deploy to your preview branch, can you give me an overview of your deploy process for preview and for production?

**Reply by charleskoehl** [reply]
We have two projects called "booking" and "signup" under one team in a Pro Vercel account. They represent separate apps within an [nx](https://nx.dev/) monorepo. Both projects use the "main" branch for production, and the "staging" branch for preview. In the project settings for both, the "Connected Git Repository" is the same monorepo. There are no deploy hooks and no custom ignored build step.

To deploy any changes at all, we just push the entire monorepo to main or staging for production and preview respectively.

**Reply by charleskoehl** [reply]
The automatically-generated vercel subdomains behave the same way.

**Reply by charleskoehl** [reply]
We did an experiment to see if the problem was just related to the auto-generated system env vars:

We added 3 versions of a new shared env var "REAL_ENV" for development, preview and production to the team environment variables, and linked all 3 to each of the 2 projects within that team. Then we output that in the script comment instead of VERCEL_ENV, like so:
```js
              // REAL_ENV = "${process.env.REAL_ENV}"
```
However, the same problem occurs. In the latest deployment of the booking project, both the  preview and production versions have "production" as the value for REAL_ENV.

**Reply by charleskoehl** [reply]
I've submitted a ticket under our Pro plan. The case number is 00108888.

**Reply by amyegan** [reply]
Thanks for sharing the ticket number! I'm confused about how your project is being built and deployed. You said that you are not using [Vercel CLI](https://vercel.com/docs/cli), deploy [hooks](https://vercel.com/docs/concepts/deployments/overview#deploy-hooks), or the [GitHub integration](https://vercel.com/docs/concepts/git/vercel-for-github) to deploy. Are you using a CI pipeline, [Vercel API](https://vercel.com/docs/rest-api#), or another method? I need to know how the project is being deployed so I can identify possible points of failure for the environment variables. 🙏

**Reply by charleskoehl** [reply]
As I wrote previously, "To deploy any changes at all, we just push the entire monorepo to main or staging for production and preview respectively." Each project has the monorepo git repo connected to it.

**Reply by amyegan** [reply]
Thank you for being patient with all of my questions! I asked the team to help look into this problem and I wanted to be sure we have the right info. Someone on the team responded to your ticket. You'll be able to share deployment information less publicly there and I think that will make detailed debugging conversations a little more comfortable. 

One idea that was suggested is monorepo caching. Have you tried [`--skip-nx-cache`](https://nx.dev/concepts/how-caching-works#skipping-cache)?

**Reply by charleskoehl** [reply]
no we'll try that. thanks!
